### ISSUE TITLE
Apparent Build Failure on Windows 10 64-bit

### ISSUE THREAD

**Issue by prdctofchem** [blue-yonder/turbodbc#262]
For all those receiving a BOOST_ROOT variable not set error, I had to set mine graphically in environment variables. Using `set BOOST_ROOT=my_path\boost_1_72_0` did not seem to work. Despite adding a BOOST_ROOT variable to my environment variables the build still fails.

I'm on Windows 10, Python 3.7.6 and I have:
- Removed all Visual Studio versions to ensure that different versions of build libraries would not conflict.
- Installed VS 2019 C++ build tools
- Downloaded boost_1_72_0
- Run .\bootstrap.bat and .\b2 from x64 Native Tools Command Prompt for VS 2019
- set BOOST_ROOT to my_path\boost_1_72_0
- pip install turbodbc
(All from the [Get Started](https://turbodbc.readthedocs.io/en/latest/pages/getting_started.html) portion of documentation)

Now it seems that many of the files setup goes looking for do not exist: many from Windows Kits, others from Visual Studio. I get an overflow of errors that include:
```
    C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.25.28610\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -Iinclude/ -Ib:\python37\lib\site-packages\pybind11\include "-IB:\Program Files\boost\boost_1_72_0\boost_1_72_0\" -Ib:\python37\include -Ib:\python37\include "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.25.28610\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt" /EHsc /Tpsrc\cpp_odbc\level3\raii_environment.cpp /Fobuild\temp.win-amd64-3.7\Release\src\cpp_odbc\level3\raii_environment.obj -DNOMINMAX
    cl : Command line warning D9024 : unrecognized source file type 'Files', object file assumed
    cl : Command line warning D9027 : source file 'Files' ignored
    cl : Command line warning D9024 : unrecognized source file type '(x86)\Microsoft', object file assumed
    cl : Command line warning D9027 : source file '(x86)\Microsoft' ignored
    cl : Command line warning D9024 : unrecognized source file type 'Visual', object file assumed
    cl : Command line warning D9027 : source file 'Visual' ignored
    cl : Command line warning D9024 : unrecognized source file type 'Studio\2019\BuildTools\VC\Tools\MSVC\14.25.28610\include -IC:\Program', object file assumed
    cl : Command line warning D9027 : source file 'Studio\2019\BuildTools\VC\Tools\MSVC\14.25.28610\include -IC:\Program' ignored
```

and the final  part of the error is:
```
    cl : Command line warning D9021 : no action performed
    C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.25.28610\bin\HostX86\x64\link.exe /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO /LIBPATH:build\lib.win-amd64-3.7 "/LIBPATH:B:\Program Files\boost\boost_1_72_0\boost_1_72_0\stage\lib" "/LIBPATH:B:\Program Files\boost\boost_1_72_0\boost_1_72_0\lib64-msvc-14.0" /LIBPATH:b:\python37\libs /LIBPATH:b:\python37\PCbuild\amd64 "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.25.28610\lib\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\um\x64" odbc32.lib /EXPORT:PyInit_turbodbc_intern build\temp.win-amd64-3.7\Release\src\cpp_odbc\column_description.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\connection.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\cpp_odbc.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\environment.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\error.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\make_environment.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\multi_value_buffer.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\statement.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\level1\api.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\level1\level1.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\level1\unixodbc_backend.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\level1\unixodbc_backend_debug.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\level2\api.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\level2\handles.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\level2\input_string_buffer.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\level2\input_u16string_buffer.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\level2\level1_connector.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\level2\level2.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\level2\string_buffer.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\level2\u16string_buffer.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\level3\level3.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\level3\raii_connection.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\level3\raii_environment.obj build\temp.win-amd64-3.7\Release\src\cpp_odbc\level3\raii_statement.obj build\temp.win-amd64-3.7\Release\src\turbodbc\buffer_size.obj build\temp.win-amd64-3.7\Release\src\turbodbc\column.obj build\temp.win-amd64-3.7\Release\src\turbodbc\command.obj build\temp.win-amd64-3.7\Release\src\turbodbc\configuration.obj build\temp.win-amd64-3.7\Release\src\turbodbc\connect.obj build\temp.win-amd64-3.7\Release\src\turbodbc\connection.obj build\temp.win-amd64-3.7\Release\src\turbodbc\cursor.obj build\temp.win-amd64-3.7\Release\src\turbodbc\description.obj build\temp.win-amd64-3.7\Release\src\turbodbc\errors.obj build\temp.win-amd64-3.7\Release\src\turbodbc\field_translator.obj build\temp.win-amd64-3.7\Release\src\turbodbc\make_description.obj build\temp.win-amd64-3.7\Release\src\turbodbc\make_field_translator.obj build\temp.win-amd64-3.7\Release\src\turbodbc\parameter.obj build\temp.win-amd64-3.7\Release\src\turbodbc\string_helpers.obj build\temp.win-amd64-3.7\Release\src\turbodbc\time_helpers.obj build\temp.win-amd64-3.7\Release\src\turbodbc\descriptions\boolean_description.obj build\temp.win-amd64-3.7\Release\src\turbodbc\descriptions\date_description.obj build\temp.win-amd64-3.7\Release\src\turbodbc\descriptions\floating_point_description.obj build\temp.win-amd64-3.7\Release\src\turbodbc\descriptions\integer_description.obj build\temp.win-amd64-3.7\Release\src\turbodbc\descriptions\string_description.obj build\temp.win-amd64-3.7\Release\src\turbodbc\descriptions\timestamp_description.obj build\temp.win-amd64-3.7\Release\src\turbodbc\descriptions\unicode_description.obj build\temp.win-amd64-3.7\Release\src\turbodbc\field_translators\boolean_translator.obj build\temp.win-amd64-3.7\Release\src\turbodbc\field_translators\date_translator.obj build\temp.win-amd64-3.7\Release\src\turbodbc\field_translators\float64_translator.obj build\temp.win-amd64-3.7\Release\src\turbodbc\field_translators\int64_translator.obj build\temp.win-amd64-3.7\Release\src\turbodbc\field_translators\string_translator.obj build\temp.win-amd64-3.7\Release\src\turbodbc\field_translators\timestamp_translator.obj build\temp.win-amd64-3.7\Release\src\turbodbc\parameter_sets\bound_parameter_set.obj build\temp.win-amd64-3.7\Release\src\turbodbc\parameter_sets\field_parameter_set.obj build\temp.win-amd64-3.7\Release\src\turbodbc\parameter_sets\set_field.obj build\temp.win-amd64-3.7\Release\src\turbodbc\result_sets\bound_result_set.obj build\temp.win-amd64-3.7\Release\src\turbodbc\result_sets\double_buffered_result_set.obj build\temp.win-amd64-3.7\Release\src\turbodbc\result_sets\field_result_set.obj build\temp.win-amd64-3.7\Release\src\turbodbc\result_sets\result_set.obj build\temp.win-amd64-3.7\Release\src\turbodbc\result_sets\row_based_result_set.obj build\temp.win-amd64-3.7\Release\src\turbodbc_python\determine_parameter_type.obj build\temp.win-amd64-3.7\Release\src\turbodbc_python\python_parameter_set.obj build\temp.win-amd64-3.7\Release\src\turbodbc_python\python_result_set.obj build\temp.win-amd64-3.7\Release\src\turbodbc_python\python_bindings\buffer_size.obj build\temp.win-amd64-3.7\Release\src\turbodbc_python\python_bindings\column_info.obj build\temp.win-amd64-3.7\Release\src\turbodbc_python\python_bindings\connect.obj build\temp.win-amd64-3.7\Release\src\turbodbc_python\python_bindings\connection.obj build\temp.win-amd64-3.7\Release\src\turbodbc_python\python_bindings\cursor.obj build\temp.win-amd64-3.7\Release\src\turbodbc_python\python_bindings\error.obj build\temp.win-amd64-3.7\Release\src\turbodbc_python\python_bindings\module.obj build\temp.win-amd64-3.7\Release\src\turbodbc_python\python_bindings\options.obj build\temp.win-amd64-3.7\Release\src\turbodbc_python\python_bindings\python_parameter_set.obj build\temp.win-amd64-3.7\Release\src\turbodbc_python\python_bindings\python_result_set.obj /OUT:build\lib.win-amd64-3.7\turbodbc_intern.cp37-win_amd64.pyd /IMPLIB:build\temp.win-amd64-3.7\Release\src\cpp_odbc\turbodbc_intern.cp37-win_amd64.lib
    LINK : fatal error LNK1181: cannot open input file 'build\temp.win-amd64-3.7\Release\src\cpp_odbc\column_description.obj'
    error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.25.28610\\bin\\HostX86\\x64\\link.exe' failed with exit status 1181
    ----------------------------------------
ERROR: Command errored out with exit status 1: 'b:\python37\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\Master\\AppData\\Local\\Temp\\pip-install-w8j707pz\\turbodbc\\setup.py'"'"'; __file__='"'"'C:\\Users\\Master\\AppData\\Local\\Temp\\pip-install-w8j707pz\\turbodbc\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\Master\AppData\Local\Temp\pip-record-5y9vcwul\install-record.txt' --single-version-externally-managed --compile --install-headers 'b:\python37\Include\turbodbc' Check the logs for full command output.
```

The error prints so fast I can't catch the first part of it. Any help is appreciated, I am chomping at the bit to test turbodbc for performance improvements in my project!

Ross

**Comment by keitherskine** [comment]
Hi @prdctofchem .  First off, I just want to make sure you're using a 64-bit version of python (run `python -VV` to check).  turbodbc requires 64-bit python.  Is that the case?
Second, just to confirm, you seem to have python installed on a "B:" drive on your system.  Is that correct?
Did you get any errors when running ` .\bootstrap.bat` and/or `.\b2.bat` ?  Generally, warnings can be ignored, but did any errors arise?  If so, what errors? (in particular the first error)
Bear in mind, when running `pip install turbodbc`, you don't have to do that from an "x64 Native Tools Command Prompt for VS 2019" prompt.  You can run it from an ordinary command prompt.  Typically, this is done within a virtual environment.
Here's what happened when I ran `pip install turbodbc` from a venv, all from an ordinary command prompt (without admin privs):
```
C:\Users\keith\Downloads\pyvenvs>py -3.7 -m venv turbodbc_262
C:\Users\keith\Downloads\pyvenvs>cd turbodbc_262
C:\Users\keith\Downloads\pyvenvs\turbodbc_262>.\Scripts\activate.bat
(turbodbc_262) C:\Users\keith\Downloads\pyvenvs\turbodbc_262>python -VV
Python 3.7.5 (tags/v3.7.5:5c02a39a0b, Oct 15 2019, 00:11:34) [MSC v.1916 64 bit (AMD64)]

(turbodbc_262) C:\Users\keith\Downloads\pyvenvs\turbodbc_262>set BOOST_ROOT=C:\Users\keith\Downloads\boost_1_72_0\boost_1_72_0
(turbodbc_262) C:\Users\keith\Downloads\pyvenvs\turbodbc_262>pip install turbodbc
Collecting turbodbc
  Downloading https://files.pythonhosted.org/packages/fc/6e/f2febb0ad9abaec5afcac0c8093fdebb2f407a418baaa2d0ef1866c0e395/turbodbc-4.0.0.tar.gz (77kB)
     |████████████████████████████████| 81kB ...
Collecting pybind11>=2.2.0 (from turbodbc)
  Using cached https://files.pythonhosted.org/packages/89/e3/d576f6f02bc75bacbc3d42494e8f1d063c95617d86648dba243c2cb3963e/pybind11-2.5.0-py2.py3-none-any.whl
Installing collected packages: pybind11, turbodbc
  Running setup.py install for turbodbc ... done
Successfully installed pybind11-2.5.0 turbodbc-4.0.0
WARNING: You are using pip version 19.2.3, however version 20.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.

(turbodbc_262) C:\Users\keith\Downloads\pyvenvs\turbodbc_262>python -c "import turbodbc; print(turbodbc.apilevel)"
2.0

(turbodbc_262) C:\Users\keith\Downloads\pyvenvs\turbodbc_262>
```

**Comment by prdctofchem** [comment]
Hey Keith, thanks for the response.

I am indeed running 64-bit python `Python 3.7.6 (tags/v3.7.6:43364a7ae0, Dec 19 2019, 00:42:30) [MSC v.1916 64 bit (AMD64)]` from a drive called B:. OS is installed on a separate SSD and I've been doing work from 2 drives for a few years.

I get no errors running bootstrap.bat or b2, it says build was successful. One thing I've wondered, though, is if I should be setting a variable for linker library paths, as mentioned at the end of the install statement:
```
The Boost C++ Libraries were successfully built!

The following directory should be added to compiler include paths:

    B:\Program Files\boost\boost_1_72_0\boost_1_72_0

The following directory should be added to linker library paths:

    B:\Program Files\boost\boost_1_72_0\boost_1_72_0\stage\lib
```

**Comment by keitherskine** [comment]
Hmm, I'm really not sure what's going on there, Ross.  I've tried doing a full boost compile and turbodbc install and it's fine on my Windows 10 Pro PC (without having to alter linker library paths).  The log for my turbodbc install is attached.  Perhaps you can compare it with your log and figure out where they diverge.  My suspicion is that it is something to do with having boost/python on your B: drive and MSVC on your C: drive, but I admit I'm guessing.

[turbodbc_install_log_20200518.txt](https://github.com/blue-yonder/turbodbc/files/4646690/turbodbc_install_log_20200518.txt)

**Comment by prdctofchem** [comment]
To give an update, I made sure `wheel` and `setuptools` packages were up to date which did get rid of a few errors I ran into. I've since realized that the environment variable is only set for the current cmd session using `set [var]=[path]`. Install is still failing, however, and the main divergence between our logs occurs on this line after `level1_connector.cpp`:

```
src\cpp_odbc\level2\level1_connector.cpp(17): fatal error C1083: Cannot open include file: 'boost/locale.hpp': No such file or directory
```
I have confirmed that that file exists at BOOST_ROOT\boost\locale.hpp, so I made sure to try running cmd as administrator in case it was a permissions issue, but that had no effect on installation. 
Full log attached.

[turbodbc_pipinstall_060720_5.txt](https://github.com/blue-yonder/turbodbc/files/4743029/turbodbc_pipinstall_060720_5.txt)

**Comment by keitherskine** [comment]
If I don't set BOOST_ROOT in my shell, I appear to get the same error you're seeing:
```
    C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.24.28314\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -Iinclude/ -Ic:\users\keith\downloads\pyvenvs\turbodbc_262_v2\lib\site-packages\pybind11\include -Ic:\users\keith\downloads\pyvenvs\turbodbc_262_v2\include -IC:\Users\keith\AppData\Local\Programs\Python\Python37\include -IC:\Users\keith\AppData\Local\Programs\Python\Python37\include "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.24.28314\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt" /EHsc /Tpsrc\cpp_odbc\level2\level1_connector.cpp /Fobuild\temp.win-amd64-3.7\Release\src\cpp_odbc\level2\level1_connector.obj -DNOMINMAX
    level1_connector.cpp
    src\cpp_odbc\level2\level1_connector.cpp(17): fatal error C1083: Cannot open include file: 'boost/locale.hpp': No such file or directory
```
Forgive me asking, but are you sure BOOST_ROOT has been set correctly in your shell?  Here are the two commands I run in an ordinary command shell, one after the other:
```
set BOOST_ROOT=C:\Users\keith\Downloads\boost_1_72_0\boost_1_72_0
pip install --force-reinstall turbodbc
```
I set BOOST_ROOT as above because the boost file "bootstrap.bat" is located on my PC here:
C:\Users\keith\Downloads\boost_1_72_0\boost_1_72_0\bootstrap.bat
In other words, if I run the command `dir "%BOOST_ROOT%\bootstrap.bat"`, I get a successful response.

By the way, I noticed you are using a VS2017 compiler rather than a VS2019 compiler, but I don't _think_ that should make a difference.

**Comment by xhochy** [comment]
> By the way, I noticed you are using a VS2017 compiler rather than a VS2019 compiler, but I don't think that should make a difference.

`turbodbc` should also work with VS2015 compiler, so this is fine.

**Comment by prdctofchem** [comment]
Finally got it!

I suspect there was more than one issue across my machines, but the last one was that when I set BOOST_ROOT I used quotes because there is a space in Program Files (`set BOOST_ROOT="C:\Program Files\boost..."). After getting through some of the other issues I realized the quotes were leading to a bad call and misinterpreted path:

```
C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.26.28801\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -Iinclude/ -Ib:\python37\lib\site-packages\pybind11\include "-I"B:\Program Files\boost\boost_1_73_0"" -Ib:\python37\include -Ib:\python37\include "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.26.28801\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt" /EHsc /Tpsrc\cpp_odbc\column_description.cpp /Fobuild\temp.win-amd64-3.7\Release\src\cpp_odbc\column_description.obj -DNOMINMAX
cl : Command line warning D9024 : unrecognized source file type 'Files\boost\boost_1_73_0', object file assumed
```

Thanks for hanging in there with me and sorry for any lost time!

**Comment by wonb168** [comment]
so complex, why must use odbc? 
use jdbc, or pure python, will easier for all users.
